doc/design/persistent_volumes_etcd_data.md

```diff
@@ -0,0 +1,125 @@
+
+#### Case A
+
+##### Event
+
+One of these events happens:
+* one etcd member process exits (processes crashes, killed etc...)
+* a k8s node executing one etcd member pod reboots and comes back before node controller pod eviction triggers
+
+##### Effect
+
+The pod will go in failed state, the etcd cluster remains quorate. Etcd-operator will remove the member from the etcd cluster and add a new member to the cluster and creates a new pod with a new name.
+
+#### Case B
+
+##### Event
+
+* a k8s node executing one etcd member pod becomes partitioned (network partitioning, goes down, etc...) and cannot talk to the api server, after some time the node controller will evict its pods.
+
+##### Effect
+
+The pod will go in an unknown/failed state, the etcd cluster remains quorate. Etcd-operator will remove the member from the etcd cluster and add a new member to the cluster and creates a new pod with a new name.
+
+#### Case C
+
+##### Event
+
+One of these events happens:
+* a majority of etcd processes exits (processes crashes, killed etc...) 
+* k8s nodes scheduling a majority of etcd members pods reboots and come back before node controller pod eviction triggers
+
+##### Effect
+
+The pods will go in failed state, the etcd cluster becomes unquorate. Etcd-operator can only restore the cluster from a backup.
+
+#### Case D
+
+##### Event
+
+* k8s nodes executing a majority of etcd members pods becomes partitioned (network partitioning, goes down, etc...) and cannot talk to the api server, after some time the node controller will evict their pods.
+
+##### Effect
+
+The pods will go in an unknown/failed state and eventually be removed from the API server, the etcd cluster becomes unquorate. Etcd-operator can only restore the cluster from a backup.
+
+
+#### Reasons
+
+Currently etcd operator handles cases A and B but when a cluster becomes unquorate (cases C and D) the current option is to restore the cluster from a backup.
+
+User may desire consistency over availability and prefer to wait for the cluster the return in a quorate state instead of restoring from a backup (that will contain old data).
+
+This proposal aims to fix cases C and D permitting to the cluster the return in a quorate state without the need to restore from a backup.
+
+## Proposed Changes
+
+#### Changes
+
+The change to be effective and cover the above failure scenario is split in two parts.
+
+### Part 1
+
+- Add to the cluster spec PodPolicy an option to enable putting etcd data inside a Persistent Volume
+- If podPolicy has persistent volumes enabled, when creating a new member also create a Persistent Volume Claim and use it as the volume source for etcd-data
+- Handle etcd data persistent volume claims cleanup:
+  - set on the PVC an owner like already done with other resources.
+  - in the reconcile loop, remove PVCs that don't have a matching etcd member in the etcd cluster and their related pod doesn't exists in the k8s API
```

Comment:
`PVC lifecycle is bound to etcd member's lifecycle. PVC is added/removed when etcd member is added/removed.`

Reply:
> remove PVCs that don't have a matching etcd member in the etcd cluster and their related pod doesn't exists in the k8s API

 We don't remove PVCs just because pods were gone. PVC lifecycle is bound to member's lifecycle.

Currently if a pod is gone (evicted maybe?), we remove its member and create a new one with a new pod. With PV support, we will keep its member, then:

- tolerate it for a while (e.g. 5 mins), waiting it to restart. Or
- after toleration period ends or pod is gone from APIServer, create a pod with the same identity and using the same PVC.

In short, we don't need to change membership with PV support. To put it more general, PVC lifecycle is bound to member's lifecycle. For example, we will remove PVC only if its member is removed in scale-down.

Reply:
Note that we don't handle failure & toleration in part 1. Just explain why it doesn't correlate to Pods.

Reply:
> PVC lifecycle is bound to etcd member's lifecycle. PVC is added/removed when etcd member is added/removed.

will add.

> We don't remove PVCs just because pods were gone. PVC lifecycle is bound to member's lifecycle.

I tough it was clear since I put an AND. the second part of the AND is just to avoid possible problems if we are going to remove a PVC before the pod it's assigned to is gone.

Reply:
OK. Sure.

Just to sync up